pyro/infer/traceenum_elbo.py

```diff
@@ -40,43 +41,58 @@ def _compute_model_costs(model_trace, guide_trace, ordering):
-                    log_factors.append(cost)
+                    log_factors.setdefault(t, []).append(cost)
                     scales.add(site["scale"])
-            if not log_factors:
-                continue
-            for u, sites_u in enum_sites.items():
+        if log_factors:
+            # TODO split log_factors into connected components wrt shared tensor dims.
+            upper = frozenset.union(*log_factors.keys())
+            for t, sites_t in enum_sites.items():
                 # TODO refine this coarse dependency ordering using time and tensor shapes.
-                if u <= t:
-                    for site in sites_u:
+                if t <= upper:
+                    for site in sites_t:
                         logprob = site["unscaled_log_prob"]
-                        log_factors.append(logprob)
+                        log_factors.setdefault(t, []).append(logprob)
                         scales.add(site["scale"])
-            # This is only correct if all enumerated things share a common subsampling scale.
+            # This is only correct if all enumerated sites share a common subsampling scale.
             # Note that we use a cheap weak comparison by id rather than tensor value, because
             # (1) it is expensive to compare tensors by value, and (2) tensors must agree not
             # only in value but at all derivatives.
             if len(scales) != 1:
                 raise ValueError("Expected all enumerated sample sites to share a common poutine.scale, "
                                  "but found {} different scales.".format(len(scales)))
-            target_shape = (broadcast_shape(*set(x.shape[enum_boundary:] for x in log_factors))
-                            if enum_boundary else ())
-            marginal_cost = logsumproductexp(log_factors, target_shape)
+
+            # Contract enumeration dimensions via sum-product algorithm (in log space).
+            shapes = set(x.shape[enum_boundary:] for xs in log_factors.values() for x in xs)
+            target_shape = broadcast_shape(*shapes) if enum_boundary else ()
+            scaled_log_factors = []
+            for t, log_factors_t in log_factors.items():
+                # Avoid double-counting due to the .sum() in the second contraction below.
+                denom = reduce(operator.mul, (f.size for f in upper - t), 1)
+                for log_factor in log_factors_t:
+                    scaled_log_factors.append(log_factor if denom == 1 else log_factor / denom)
+            marginal_cost = logsumproductexp(scaled_log_factors, target_shape)
+
+            # Contract iarange dimensions via product (in log space).
+            lower = frozenset.intersection(*log_factors.keys())
+            for frame in upper - lower:
+                marginal_cost = marginal_cost.sum(frame.dim, keepdim=True)
+            while marginal_cost.dim() and marginal_cost.shape[0] == 1:
+                marginal_cost.squeeze_(0)
```

Comment:
if i followed this correctly, youll have still singleton dims within `marginal_cost` (because of the previous `sum`)?

Reply:
Yes, this is typical when `iaranges` skip dimensions, e.g. when `vectorize_particles` adds an `iarange` to the left of all other `iarange`s, and say we're only enumerating over a global variable. For example in the following we may want to compare multiple models by sampling `which` model we believe is useful.
```py
which = pyro.sample("which", Categorical(...))
with iarange("data", len(data)):
    pyro.sample("obs", Normal(locs[which], scales[which]), obs=data)
```
If we run this with `num_particles=10, vectorize_particles=True`, then the final `marginal_cost` will have shape `(num_particles, 1)`.